lib/scout_apm/instant/middleware.rb

```diff
@@ -45,66 +50,185 @@ def initialize(app)
-                end
-              end
-            else
-              ScoutApm::Agent.instance.logger.debug("DevTrace: in middleware, dev_trace is active, and response has a body, but no trace was found. Path=#{path}; ContentType=#{content_type}")
-              [status, headers, response]
-            end
-          else
-            # don't log anything here - this is the path for all assets served in development, and the log would get noisy
-            [status, headers, response]
-          end
+        rack_response = @app.call(env)
+        begin
+          DevTraceResponseManipulator.new(env, rack_response).call
+        rescue => e
+          ScoutApm::Agent.instance.logger.debug("DevTrace: Raised an exception: #{e.message}, #{e.backtrace}")
+          rack_response
+        end
+      end
+    end
+
+    class DevTraceResponseManipulator
+      attr_reader :rack_response
+      attr_reader :rack_status, :rack_headers, :rack_body
+      attr_reader :env
+
+      def initialize(env, rack_response)
+        @env = env
+        @rack_response = rack_response
+
+        @rack_status = rack_response[0]
+        @rack_headers = rack_response[1]
+        @rack_body = rack_response[2]
+      end
+
+      def call
+        return rack_response unless preconditions_met?
+
+        if ajax_request?
+          ScoutApm::Agent.instance.logger.debug("DevTrace: in middleware, dev_trace is active, and response has a body. This is either AJAX or JSON. Path=#{path}; ContentType=#{content_type}")
+          adjust_ajax_header
         else
-          ScoutApm::Agent.instance.logger.debug("DevTrace: isn't activated via config. Try: SCOUT_DEV_TRACE=true rails server")
-          [status, headers, response]
+          adjust_html_response
+        end
+
+        rebuild_rack_response
+      end
+
+      ###########################
+      #  Precondition checking  #
+      ###########################
+
+      def preconditions_met?
+        if dev_trace_disabled?
+          logger.debug("DevTrace: isn't activated via config. Try: SCOUT_DEV_TRACE=true rails server")
+          return false
         end
+
+        # Don't attempt to instrument assets.
+        # Don't log this case, since it would be very noisy
+        logger.debug("DevTrace: dev asset ignored") and return false if development_asset?
+
+        # If we didn't have a tracked_request object, or we explicitly ignored
+        # this request, don't do any work.
+        logger.debug("DevTrace: no tracked request") and return false if tracked_request.nil? || tracked_request.ignoring_request?
+
+        # If we didn't get a trace, we can't show a trace...
+        if trace.nil?
+          logger.debug("DevTrace: in middleware, dev_trace is active, and response has a body, but no trace was found. Path=#{path}; ContentType=#{content_type}")
+          return false
+        end
+
+        true
+      end
+
+      def dev_trace_disabled?
+        ! ScoutApm::Agent.instance.config.value('dev_trace')
```

Comment:
This middleware is only included in Rails `development` mode, none of this code should ever run in production.

In dev mode, I'm not too worried, this is checked really early in the sequence here. But we could avoid making this class at all I suppose, by checking this config setting in the Middleware, instead of waiting until we created the class and `.call`ed it.